src/tokenizer.js

```diff
@@ -328,31 +372,55 @@ class Tokenizer {
                 }
                 // chrome doens't show contents of text nodes with very long text
                 while (lastIndex < line.length) {
-                    if (token.type)
-                        tokens.push(token);
+                    if (token.type) tokens.push(token);
                     token = {
                         value: line.substring(lastIndex, lastIndex += 500),
-                        type: "overflow"
+                        type: currentState.get("overflow")
                     };
                 }
-                currentState = "start";
+                currentState = this.rootScope.get("start");
                 stack = [];
                 break;
             }
         }
 
-        if (token.type)
+        if (token.type) tokens.push(token);
+
+        if (!tokens.length || tokens[tokens.length - 1].type.parent !== currentState) {
+            token = {
+                value: "",
+                type: currentState.get("empty")
+            };
             tokens.push(token);
-        
-        if (stack.length > 1) {
-            if (stack[0] !== currentState)
-                stack.unshift("#tmp", currentState);
         }
+
         return {
-            tokens : tokens,
-            state : stack.length ? stack : currentState
+            tokens: tokens,
+            state: currentState
         };
     }
+
+    /**
+     * Retrieves the scope for the given stack and current state.
+     *
+     * @param {string[]} stack - The stack of scopes.
+     * @param {string} currentState - The current state.
+     * @returns {Scope} The scope for the given stack and current state.
+     */
+    fromStack(stack, currentState) {
```

Comment:
fromStack, toStack should be in one place, maybe keep them as unexported functions in this file if they are not used anywhere else.